### ISSUE TITLE
Can't type after resuming from suspend

### ISSUE THREAD

**Issue by lucasreis1** [hyprwm/hyprlock#101]
I'm using `hypridle` to trigger `hyprlock` before suspending. After resuming, I can move my mouse and the clock updates, which signals that `hypridle` is working fine, but keyboard input won't work in the password box.

`hypridle.conf`:
```
general {
  before_sleep_cmd = WAYLAND_DEBUG=1 ~/.config/hypr/scripts/lock.sh &> /tmp/hyprlock_out.txt
  after_sleep_cmd = ~/.config/hypr/scripts/wp_init.sh -r
}

listener {
  timeout = 600
  on-timeout = ~/.config/hypr/scripts/lock.sh --dimfirst
}

listener {
  timeout = 720
  on-timeout = hyprctl dispatch dpms off
}
```

`lock.sh`:
```
#/bin/bash

WALLPAPER_DIR=~/Pictures/wallpapers/blurred
CURRENT_WALL=$(wpg -c)


EXTENSION="${CURRENT_WALL##*.}"
# for gifs, replace fframe with the gif.
# also, remember they are stored in the subfolder "animated"
if [[ "$EXTENSION" =~ fframe.*$ ]]; then
  CURRENT_WALL=${CURRENT_WALL%.*}.png
fi

IMG="$WALLPAPER_DIR/$CURRENT_WALL"

if [ ! -f $IMG ];
then
  IMG=$(ls $WALLPAPER_DIR/* | shuf | head -n 1)
fi
cp $IMG /tmp/hyprlock.png
#swaylock -f -i $IMG
if [ "$1" == "--dimfirst" ]
then
  chayang -d 10 && hyprlock 
else
  hyprlock
fi
```

`hyprland.log`:
```
[LOG] Instance Signature: 29cdd7de1f8d99462915540569fbe54f10f609f4_1709181372
[LOG] Hyprland PID: 396621
[LOG] ===== SYSTEM INFO: =====
[LOG] System name: Linux
[LOG] Node name: lucas
[LOG] Release: 6.7.6-arch1-1
[LOG] Version: #1 SMP PREEMPT_DYNAMIC Fri, 23 Feb 2024 16:31:48 +0000


[LOG] GPU information:
01:00.0 VGA compatible controller [0300]: NVIDIA Corporation GA104 [GeForce RTX 3070 Lite Hash Rate] [10de:2488] (rev a1) (prog-if 00 [VGA controller])


[WARN] Warning: you're using an NVIDIA GPU. Make sure you follow the instructions on the wiki if anything is amiss.

[LOG] os-release:
NAME="Arch Linux"
PRETTY_NAME="Arch Linux"
ID=arch
BUILD_ID=rolling
ANSI_COLOR="38;2;23;147;209"
HOME_URL="https://archlinux.org/"
DOCUMENTATION_URL="https://wiki.archlinux.org/"
SUPPORT_URL="https://bbs.archlinux.org/"
BUG_REPORT_URL="https://gitlab.archlinux.org/groups/archlinux/-/issues"
PRIVACY_POLICY_URL="https://terms.archlinux.org/docs/privacy-policy/"
LOGO=archlinux-logo

[LOG] ========================



[INFO] If you are crashing, or encounter any bugs, please consult https://wiki.hyprland.org/Crashes-and-Bugs/


[LOG] 
Current splash: Take on me, take me on...


[LOG] Creating the HookSystem!
[LOG] Creating the KeybindManager!
[LOG] [hookSystem] New hook event registered: configReloaded
[LOG] Creating the AnimationManager!
[LOG] Created a bezier curve, baked 255 points, mem usage: 4.08kB, time to bake: 25.07µs. Estimated average calc time: 0.14µs.
[LOG] Creating the ConfigManager!
[LOG] NOTE: further logs to stdout / logfile are disabled by default. Use debug:disable_logs and debug:enable_stdout_logs to override this.
[LOG] Created a bezier curve, baked 255 points, mem usage: 4.08kB, time to bake: 10.90µs. Estimated average calc time: 0.14µs.
[LOG] Using config: /home/lucas/.config/hypr/hyprland.conf
```

[hyprlock_out.txt](https://github.com/hyprwm/hyprlock/files/14450155/hyprlock_out.txt)

**Comment by LeGmask** [comment]
I've the same issue, after suspend I cannot type my password anymore...

**Comment by ArvineKwok** [comment]
same here

**Comment by MathisP75** [comment]
Same issue here

**Comment by harshkhandeparkar** [comment]
I would like to add that for me, the textbox types 3 dots for each character typed after waking from suspend.

**Comment by FearlessSpiff** [comment]
> I would like to add that for me, the textbox types 3 dots for each character typed after waking from suspend.

this is fixed. build from source...

**Comment by dtk9289** [comment]
bump

**Comment by lucasreis1** [comment]
I'm sorry but why is this closed as completed? I'm still experiencing the issue on latest git for hyprland, hypridle and hyprlock.

**Comment by vaxerski** [comment]
I assumed "this is fixed" and 3 likes meant this is fixed. Are you sure you're on latest git? no duplicate binaries?

**Comment by sefodopo** [comment]
I can also confirm that I had this issue on the latest git this morning when I woke up to wake up my computer from sleep. It wouldn't take in any input, also if I remember correctly, the mouse cursor was also showing which isn't supposed to be showing.

However, after this issue was closed, I looked around at other issues and ran across #76. In that issue it was mentioned that the crashing only occurs when running `systemctl suspend` from a terminal and not when ran by hypridle. So far today, after making that change (and updating to the latest commit this morning b4a1c8c, I haven't had this issue. But I also haven't went to sleep overnight again.
I will update hyprlock again before bed tonight to the latest git and test this again.

I think the "this is fixed" was referring to a side tangent of "the textbox types 3 dots for each character typed after waking from suspend" which isn't what it looks like the first post was talking about.

**Comment by lucasreis1** [comment]
> I assumed "this is fixed" and 3 likes meant this is fixed. Are you sure you're on latest git? no duplicate binaries?

I'm pretty sure he was replying to the person who pointed to another adjacent issue (typing three dots for each character). The original problem from suspending is still present here.

**Comment by lucasreis1** [comment]
> In that issue it was mentioned that the crashing only occurs when running `systemctl suspend` from a terminal and not when ran by hypridle

Can you clarify what running "suspending by hypridle" means? The only way I was aware of running suspend without writing to system files is directly from systemd.

**Comment by sefodopo** [comment]
So if I run `systemctl suspend` from kitty, then go to sleep while using hypridle with `lock_cmd = pidof hyprlock || hyprlock` and `before_sleep_cmd = loginctl lock-session`, Hyprlock won't let me input any text in the morning when I wake up and wake my computer.

On the other hand if I add a listener in hypridle to `on-timeout = systemctl suspend` and wait for my computer to go to sleep (instead of manually running `systemctl suspend`, I was unable to recreate this bug. Hyprlock worked correctly upon waking up. Also, hyprlock seemed to work okay when I did `hyprctl keyword exec systemctl suspend`.  Granted, I haven't tested any of these other methods overnight. It may take time to cause the issue, or the latest commits late last night or early this morning fixed this bug.

I also did a quick test earlier today running `systemctl suspend` from kitty, and then immediately waking the computer up within a minute, and hyprlock worked correctly. So either the latest commits fixed it, or it takes all night for an issue to appear.

**Comment by FearlessSpiff** [comment]
> > I assumed "this is fixed" and 3 likes meant this is fixed. Are you sure you're on latest git? no duplicate binaries?
> 
> I'm pretty sure he was replying to the person who pointed to another adjacent issue (typing three dots for each character). The original problem from suspending is still present here.

Yes. The three dots is fixed. The other problem I don't know. Never a good idea to add other issues in an issue.

**Comment by hakan-demirli** [comment]
I can replicate what @sefodopo described, on NixOS on e9a57f0dae1dfa02240040f5975d048175440c2c commit.

**Comment by PaideiaDilemma** [comment]
So I suspect, that hypridle is running multiple times for you guys. That way I was able to reproduce it (happened to myself by accident, because in nix you don't have to exec it since hypridle starts when `services.hypridle.enable = true`).

That would cause hyprlock to be launched twice and one instance would receive the `finished` event which would lock it up.

**Comment by PaideiaDilemma** [comment]
You could try this patch and see if you are still able to repro.
[patch.txt](https://github.com/hyprwm/hyprlock/files/14550574/patch.txt)

**Comment by PaideiaDilemma** [comment]
If your `before_sleep` script gets called twice shouldn't hyprlock get called twice as well?

**Comment by Cody-W-Tucker** [comment]
> If I could take a wild guess, I'd say nvidia is to blame here.

Non nvidia user here with the same problem. But other NixOs users have a similar problem with swaylock and swayidle. https://github.com/swaywm/swaylock/issues/344

> So I suspect, that hypridle is running multiple times for you guys. That way I was able to reproduce it (happened to myself by accident, because in nix you don't have to exec it since hypridle starts when `services.hypridle.enable = true`).

I searched NixOs options unstable and couldn't find it. Where is that service?

**Comment by PaideiaDilemma** [comment]
> Where is that service?

The option is in the hm-module contained in this repo :)

As described in #190 I ran into this issue.
Not sure if that fixes it but we will see.

But it is good to check that there aren't multiple hyprlock instances running. Can lead to similar symptoms.

**Comment by Cody-W-Tucker** [comment]
> > Where is that service?
> 
> The option is in the hm-module contained in this repo :)

That's really cool, how can I use it? Is it with a flake?

**Comment by fufexan** [comment]
> > > Where is that service?
> > 
> > 
> > The option is in the hm-module contained in this repo :)
> 
> That's really cool, how can I use it? Is it with a flake?

Yup. You add `inputs.hyprlock.url = "github:hyprwm/hyprlock";` to your flake, then add `inputs.hyprlock.homeManagerModules.default` to your HM's `imports` list.

**Comment by PaideiaDilemma** [comment]
happened again to me, but I don't see anything wrong in hyprlock when it happens.
probably a wlr issue.

**Comment by MathisP75** [comment]
I am still facing this issue on the latest git

**Comment by vaxerski** [comment]
[patch.txt](https://github.com/hyprwm/hyprlock/files/14655187/patch.txt)
try this

**Comment by MathisP75** [comment]
`error: src/managers/input/InputManager.cpp: No such file or directory`
Where even is `InputManager.cpp`?

**Comment by vaxerski** [comment]
thats hyprland

**Comment by RoccoRakete** [comment]
I can confirm, I'm experiencing this issue on nixOS too.

**Comment by lucasreis1** [comment]
> I applied the patch to the latest git of Hyprland and the issue still persists.

Same here. Applied to commit `356414639fc06ace7045236ec41d673342981057`, no changes.

**Comment by PaideiaDilemma** [comment]
And are you guys sure that hyprlock has not been launched multiple times?

**Comment by RoccoRakete** [comment]
> And are you guys sure that hyprlock has not been launched multiple times?

Pretty sure, since it happens only sometimes. Manly when the laptop is on sleep, locked, and I unplug the monitor, before I wake the laptop up.

**Comment by PaideiaDilemma** [comment]
Ok yeah that sounds like hyprlock loosing focus. Have not experienced that since https://github.com/hyprwm/Hyprland/pull/5117

**Comment by lucasreis1** [comment]
Here's some logs from Hyprland (latest git) after I replicated the issue again:

[log.txt](https://github.com/hyprwm/hyprlock/files/14779766/log.txt)

Oddly enough it seems like hyprlock is spawning after resume even if I specify as the `before_sleep_cmd`

**Comment by SebastianStork** [comment]
I also have this issue. Doing `systemctl suspend` either from the terminal or a keybind results in not being able to log back in afterwards. Suspend through hypridle has no issues. 

I am 100% sure that the issue is from hyprlock losing focus. When I focus kitty before suspending and then wake up, I am able to run commands in the still open terminal (I can see them in the bash history later).

I am on the latest git. I use NixOS and nvidia.

**Comment by Omrigan** [comment]
Also have this. Launching hyprlock with `pidof hyprlock || hyprlock`. `hypridle` config is:
```
general {
    before_sleep_cmd = lock.sh
    ignore_dbus_inhibit = false             # whether to ignore dbus-sent idle-inhibit requests (used by e.g. firefox or steam)
}

listener {
    timeout = 300                            # in seconds
    on-timeout = idlelock.sh
}
```

When hyprlock is triggered by timeout or launched directly, this doesn't happen.

Any suggestions what to do or workarounds?

**Comment by Omrigan** [comment]
Actually, just checked:
 - hypridle + swaylock: same issue
 - swayidle + swaylock: no issue
 - swayidle + hyprlock: no issue

So I suspect it is a hypridle thing :upside_down_face:

**Comment by endeavour** [comment]
Also seeing this issue although not consistently. Perhaps it's a race condition?

**Comment by djvs** [comment]
I have the issue with swayidle + hyprlock, on a non-systemd setup (elogind).  Quickly reproducible by power cycling my monitor in a lock state.  Also, power cycling the monitor in an unlocked state changes the workspace, which I can't help but feel is related.

**Comment by BRS5672023** [comment]
I find it rather peculiar that there's no issue unlocking hyprlock if I lock the session manually, say, by set `bind=$mainMod,U,exec,hyprlock` in hyprland.conf. But it is quite frequent that hyprlock does not read my password input if hyprlock is invoked by hypridle. For reference, the following is my hypridle.conf
```
general {
	ignore_dbus_inhibit = false
	lock_cmd = pidof hyprlock || hyprlock
	before_sleep_cmd = loginctl lock-session
	after_sleep_cmd = hyprctl dispatch dpms on
}

listener {
	timeout = 600
	on-timeout = loginctl lock-session
}

listener {
	timeout = 630
	on-timeout = hyprctl dispatch dpms off
	on-resume = hyprctl dispatch dpms on
}

listener {
	timeout = 30
	on-timeout = pidof hyprlock && hyprctl dispatch dpms off
	on-resume = hyprctl dispatch dpms on
}
```
And this issue is not present if I enter the password right after the locking timeout, but only after certain amount of time passed from this timeout (such behavior presents both on my laptop and desktop pc), this very same issue also appears when I use swaylock instead...

**Comment by MoltenB** [comment]
I am having the same issue. Works fine after hibernation, not using suspend.

**Comment by SebastianStork** [comment]
I did some more experimentation. When I have 
```
general {
  lock_cmd = pidof hyprlock || hyprlock
  before_sleep_cmd = loginctl lock-session
  after_sleep_cmd = hyprctl dispatch dpms on
}
```
and do a `systemctl suspend`, I get an unresponsive black screen after wake. When I do a `loginctl lock-session && systemctl suspend` without the `before_sleep_cmd` I get the same result. `loginctl lock-session && sleep 1 && systemctl suspend` works without a problem. Sadly this can't be transferred to `before_sleep_cmd = loginctl lock-session && sleep 1` as hypridle doesn't wait for the command to finish.

It seems to me like my issue is caused by not giving hyprlock time to launch before suspending. Any delay between the two fixes the issue. 

But my current issue is different than the one I reported back in april. Back then I just couldn't focus hyprlock to log back in after a suspend, everything else was fine. Now I get a black screen and my pc becomes completely unresponsive. The caps lock light doesn't work but audio that played before the suspend keeps playing.

**Comment by PaideiaDilemma** [comment]
@SebastianStork can you try this patch for hyprIDLE?

[patch.txt](https://github.com/user-attachments/files/16311838/patch.txt)

**Comment by SebastianStork** [comment]
> @SebastianStork can you try this patch for hyprIDLE?
> 
> [patch.txt](https://github.com/user-attachments/files/16311838/patch.txt)

Is there an easy way to apply the patch on NixOS?

**Comment by PaideiaDilemma** [comment]
1. clone hypridle
2. apply the patch with `wget https://github.com/user-attachments/files/16311838/patch.txt && git apply patch.txt`
3. `nix run .`

Maybe kill your running instance of hypridle first as well.

**Comment by SebastianStork** [comment]
I get the same result result with the patch: a black screen after wake, all keybinds are unresponsive, caps lock light is unresponsive, audio keeps playing

**Comment by PaideiaDilemma** [comment]
Ok and this?
[patch2.txt](https://github.com/user-attachments/files/16319637/patch2.txt)

Please also report if you see hyprlock start up, before the system suspends and monitors go dark.
Also check if using `before_sleep_cmd = pidof hyprlock || hyprlock` instead of `loginctl lock-session` works.

**Comment by SebastianStork** [comment]
`before_sleep_cmd = loginctl lock-session` doesn't work. It goes straight to black and I experience the same issues I described above.
With `before_sleep_cmd = pidof hyprlock || hyprlock` it works. The lock screen shows up and after ~4 seconds the pc goes to sleep.

**Comment by SebastianStork** [comment]
Strangely, if I first lock the screen using `loginctl lock-session` and then suspend using a keybind, I end up with two lock screens, one from the initial screen lock and another from the suspend action. However, this issue doesn't occur if I lock the screen with `pidof hyprlock || hyprlock` before suspending.

There seems to be an issue with `lock_cmd = pidof hyprlock || hyprlock`.

**Comment by PaideiaDilemma** [comment]
Thanks for testing @SebastianStork.

I am a bit surprised that I cannot reproduce that. But it seem like going through dbus via `loginctl lock-session` causes hyprlock to be launched too late or smth. I think for now you have a workaround by using `before_sleep_cmd = pidof hyprlock || hyprlock` for now.

> Strangely, if I first lock the screen using loginctl lock-session and then suspend using a keybind, I end up with two lock screens, one from the initial screen lock and another from the suspend action.

Obviously, since first lock is triggered by `loginctl lock-session` and hypridles `lock_cmd` and the  second one from the `before_sleep_cmd`.

**Comment by gk5885** [comment]
I  have been having this problem as well.  Following the advice above, I changed my config (via home manager) to the following, but I'm still having the issue described in the bug.

```
services.hypridle = {
    enable = true;
    settings = {
      general = {
        lock_cmd = "pidof hyprlock || hyprlock";
        before_sleep_cmd = "pidof hyprlock || hyprlock";
        after_sleep_cmd = "hyprctl dispatch dpms on";
      };

      listener = [
        {
          timeout = 300; # 5 min
          on-timeout = "pidof hyprlock || hyprlock";
        }
        {
          timeout = 600; # 10 min
          on-timeout = "systemctl suspend";
        }
      ];
    };
  };
```
I'm on hyprlock 0.5.0 and hypridle 0.1.5 from nixpkgs.  I have an AMD graphics card, so it doesn't appear to be a nvidia driver thing.

**Comment by leen2233** [comment]
still experiencing same issue on my laptop. sometimes works sometimes doesn't. i don't suspend with terminal, having issue when hypridle locks itself. hyprland is at the last version, newly installed.

**Comment by ReFleXzZ** [comment]
Having the same issue. Cannot even change the tty! Is this annoying.. also always need to restart the computer. Interestingly though, it's not happening always.

**Comment by adlrwbr** [comment]
Same here. When I wake from suspend, hyprlock freezes waiting to validate my password. Then, when I switch to tty3 and kill hypridle/lock, I am unable to type at all. Anyone have a workaround that doesn't require a reboot?

**Comment by voneiden** [comment]
If you can access tty then having this in a bash script can be convenient. I experience a side effect with this though hyprlock will re-lock the computer some time later spontaneously (once). 

```
hyprctl --instance 0 'keyword misc:allow_session_lock_restore 1'
hyprctl --instance 0 'dispatch exec hyprlock'
``` 

I've also recently noticed that if I wait long enough, perhaps sleeping and waking the computer a few times and generating some input events, hyprlock eventually starts to respond to keyboard input again. But I'm not sure if this works every time or if I've just been lucky.

**Comment by PaideiaDilemma** [comment]
@adlrwbr @ReFleXzZ 

I am curious: Do you use focus follows mouse, or do you use manual focus in Hyprland?
Also Nvidia users? Hyprlock version? Hyprland version?

I would generally preferred if you open a new issue, because this one is very broad.
It can have different causes why you aren't able to type after suspend. If you can't switch ttys any more, Hyprland is almost certainly also involved.

I will close this, to encourage people to open a new issue.

**Comment by tomaspinho** [comment]
Reproduced this cleanly on a live system (not a lab repro) with `WAYLAND_DEBUG=1 hyprlock -v` piped to a log across the actual suspend/resume/VT-switch, plus Hyprland's own backend log for the same window. Sharing in case the concrete evidence is useful — happy to share full logs if wanted.

### Environment

```
Hyprland 0.56.2 (commit efb50993, tag v0.56.2)
hyprlock 0.9.6
hypridle 0.1.8
aquamarine 0.15.0
nvidia-utils 610.57.04 (NVIDIA open kernel module)
Kernel 7.2.4-arch1-2
Arch Linux
```

`hypridle.conf` uses the standard wiki-example pattern:
```
general {
  lock_cmd = pidof hyprlock || hyprlock
  before_sleep_cmd = loginctl lock-session
  after_sleep_cmd = hyprctl dispatch dpms on
}
listener { timeout = 330; on-timeout = loginctl lock-session }
listener { timeout = 360; on-timeout = hyprctl dispatch dpms off; on-resume = hyprctl dispatch dpms on }
listener { timeout = 900; on-timeout = systemctl suspend }
```

### What the trace shows

Right at resume, `hyprlock` logs:

```
[ERR]: Invalid key down event (stray release event?)
```

This lines up exactly with the `WAYLAND_DEBUG` trace: the compositor tears down the pre-suspend `wl_keyboard` object and creates a fresh one at resume (seat capabilities drop to `0` then re-announce), and the *press* half of the first post-resume keystroke is lost in that handoff — landing on/discarded from the stale object — while the matching *release* arrives on the new one. `hyprlock` correctly flags this internally, but the session never recovers afterward: further correctly-typed passwords over the following ~14s (confirmed delivered at the protocol level — full `wl_keyboard` press/release streams) did not unlock it, and the same process stayed wedged through two more manual attempts several minutes apart.

### A likely contributing factor: VT switches fully re-enumerate every input device here

Hyprland's own log (`aquamarine` backend) shows that a VT switch — not just a suspend/resume — triggers a *complete* teardown and re-scan of every input device on this system:

```
ERR from aquamarine ]: drm: Session inactive
DEBUG from aquamarine ]: [libseat] [libseat/backend/logind.c:355] Enabling seat
DEBUG from aquamarine ]: [libinput] event2  - <device>: is tagged by udev as: ...
... (repeated for all ~15 connected input devices — 3 keyboards' worth of HID endpoints, a YubiKey, etc.)
DEBUG from aquamarine ]: drm: Restoring after VT switch
```

This is a much larger, slower race window than a lightweight resume, and is likely what actually drives the seat-capability churn behind the keystroke-loss above. Machines with a suspend flow that goes through a VT-inactive transition (this one apparently does) seem to inherit this same exposure.

### Once wedged, recovery needed a full Hyprland restart

For anyone hitting a similarly stuck session: killing and respawning `hyprlock` did **not** recover it. Three attempts (0.3s, 3s, 45s gaps between kill and respawn) all failed identically with:

```
DEBUG: Locking session
DEBUG: onLockFinished called. Seems we got yeeten. Is another lockscreen running?
```

with only one `hyprlock` process ever running (no duplicates). This suggests Hyprland's own session-lock state stayed stuck considering the original (dead) client's lock still active, independent of timing. `loginctl unlock-session` also had no effect (it's a D-Bus advisory signal, not authoritative over the Wayland-level lock). Only a full compositor restart (`hyprctl dispatch exit` on stock Hyprland) cleared it and allowed a normal fresh lock/unlock cycle afterward.

Happy to provide the full `WAYLAND_DEBUG` + Hyprland logs if that'd help narrow this down further.
